File: src/games/lottery.ts
import { nanoid } from 'nanoid'
import { randomInclusive } from '../util/random'

export interface LotteryProps<Player> {
  id?: string
  creator: Player
  bet: number
  players?: Player[]
  startTime?: Date
}

export class Lottery<Player> {
  public readonly id: string
  public readonly bet: number
  public readonly creator: Player
  private _players: Set<Player>
  private _startTime?: Date

  constructor({ id, creator, bet, players, startTime }: LotteryProps<Player>) {
    if (bet <= 0) {
      throw new Error('bet cannot be less than or equal to 0')
    }
    if (!Number.isInteger(bet) || Number.isNaN(bet)) {
      throw new Error('bet must be an integer')
    }

    this.id = id ?? nanoid()
    this.bet = bet

    this.creator = creator
    this._players = new Set<Player>()
    if (players) {
      players.forEach((p) => this.addPlayer(p))
    }
    if (startTime) {
      this._startTime = startTime
    }
  }

  get startTime(): Date | undefined {
    return this._startTime
  }

  get buyIn(): number {
    return this.bet
  }

  get potSize(): number {
    // The pot size is the amount of the win, while the payout doesn't include the buy-in
    // since the offset returned must account for the player not having actually lost the rep yet
    // So unlike the payout it includes the bet times all players
    return this.bet * this._players.size
  }

  get players(): Player[] {
    return [...this._players.values()]
  }

  start(): Date {
    if (this._startTime) return this._startTime
    this._startTime = new Date()
    return this._startTime
  }

  addPlayer(player: Player): void {
    this._players.add(player)
  }

  removePlayer(player: Player): void {
    this._players.delete(player)
  }

  canFinish(): boolean {
    return this._players.size > 1
  }

  /** End the lottery and get an array of players and payout amounts. Negative payouts indicate loss */
  finish(): { winner: Player; payouts: [Player, number][] } {
    const players = Array.from(this._players.values())
    const winner = players[randomInclusive(0, this._players.size - 1)]

    const payouts = players.map<[Player, number]>((player) => [
      player,
      // If the player is the loser they lose the pot size
      // otherwise the player receives the bet
      player === winner ? this.getPayout() : this.bet,
    ])

    if (!winner) {
      console.log('debug', winner, players)
    }

    return { winner, payouts }
  }

  getPayout() {
    // In a lottery everyone pays the winner, but the winner does not pay themself
    // Since money is not collected unless it is lost
    return Math.abs(this.bet) * (this._players.size - 1)
  }

  toJSON() {
    return {
      id: this.id,
      creator: this.creator,
      bet: this.bet,
      players: this.players,
      startTime: this._startTime,
    }
  }
}

File: src/games/lottery.rs
use nanoid::nanoid;
use serde::{Deserialize, Serialize};

use crate::discord::types::GuildMember;
use crate::util::random::random_inclusive;

/// Serializable player type for lottery persistence in Firestore.
/// Uses String for joined_at (not DateTime) to match existing Firestore data format.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DbPlayer {
    pub id: String,
    pub guild_id: String,
    pub username: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub joined_at: Option<String>,
}

impl From<&GuildMember> for DbPlayer {
    fn from(m: &GuildMember) -> Self {
        DbPlayer {
            id: m.id.clone(),
            guild_id: m.guild_id.clone(),
            username: m.username.clone(),
            joined_at: m.joined_at.map(|d| d.to_rfc3339()),
        }
    }
}

impl From<&DbPlayer> for GuildMember {
    fn from(p: &DbPlayer) -> Self {
        GuildMember {
            id: p.id.clone(),
            guild_id: p.guild_id.clone(),
            username: p.username.clone(),
            joined_at: p.joined_at.as_ref().and_then(|s| s.parse().ok()),
        }
    }
}

/// A generic lottery/gambling game. Players join by paying a bet,
/// and a random winner takes the pot.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Lottery<Player: Clone + PartialEq> {
    pub id: String,
    pub bet: i64,
    pub creator: Player,
    pub players: Vec<Player>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub start_time: Option<String>,
    #[serde(default)]
    pub closed: bool,
}

pub struct LotteryResult<Player> {
    pub winner: Player,
}

impl<Player: Clone + PartialEq> Lottery<Player> {
    pub fn new(creator: Player, bet: i64) -> anyhow::Result<Self> {
        if bet <= 0 {
            anyhow::bail!("bet cannot be less than or equal to 0");
        }

        Ok(Self {
            id: nanoid!(),
            bet,
            creator,
            players: Vec::new(),
            start_time: None,
            closed: false,
        })
    }

    pub fn buy_in(&self) -> i64 {
        self.bet
    }

    /// The pot is bet * number of players (the total money at stake).
    pub fn pot_size(&self) -> i64 {
        self.bet * self.players.len() as i64
    }

    /// Record the start time (idempotent).
    pub fn start(&mut self) -> String {
        if let Some(ref t) = self.start_time {
            return t.clone();
        }
        let now = chrono::Utc::now().to_rfc3339();
        self.start_time = Some(now.clone());
        now
    }

    pub fn add_player(&mut self, player: Player) {
        self.players.push(player);
    }

    /// A lottery can finish only with more than 1 player.
    pub fn can_finish(&self) -> bool {
        self.players.len() > 1
    }

    pub fn is_closed(&self) -> bool {
        self.closed
    }

    /// Finish the lottery: pick a random winner.
    pub fn finish(&self) -> LotteryResult<Player> {
        let idx = random_inclusive(0, self.players.len() as i64 - 1) as usize;
        let winner = self.players[idx].clone();
        LotteryResult { winner }
    }
}
